testing/scenario_app/android/app/src/main/java/dev/flutter/scenarios/ExternalTextureFlutterActivity.java

```diff
@@ -0,0 +1,463 @@
+        }
+      case "media":
+        if (VERSION.SDK_INT >= VERSION_CODES.LOLLIPOP) {
+          return new MediaSurfaceRenderer(this::createMediaExtractor, extras.getInt("rotation", 0));
+        } else {
+          throw new RuntimeException("MediaSurfaceRenderer not supported");
+        }
+      case "canvas":
+      default:
+        return new CanvasSurfaceRenderer();
+    }
+  }
+
+  private MediaExtractor createMediaExtractor() {
+    // Sample Video generated with FFMPEG.
+    // ffmpeg -loop 1 -i ~/engine/src/flutter/lib/ui/fixtures/DashInNooglerHat.jpg -c:v libx264
+    // -profile:v main -level:v 5.2 -t 1 -vf scale=192:256 -b:v 1M sample.mp4
+    try {
+      MediaExtractor extractor = new MediaExtractor();
+      AssetFileDescriptor afd = getAssets().openFd("sample.mp4");
+      extractor.setDataSource(afd.getFileDescriptor(), afd.getStartOffset(), afd.getLength());
+      return extractor;
+    } catch (IOException e) {
+      e.printStackTrace();
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public void onPause() {
+    surfaceViewRenderer.destroy();
+    flutterRenderer.destroy();
+    surfaceTextureEntry.release();
+    super.onPause();
+  }
+
+  @Override
+  public void onFlutterUiDisplayed() {
+    surfaceTextureEntry = getFlutterEngine().getRenderer().createSurfaceTexture();
+    SurfaceTexture surfaceTexture = surfaceTextureEntry.surfaceTexture();
+    surfaceTexture.setDefaultBufferSize(SURFACE_WIDTH, SURFACE_HEIGHT);
+    flutterRenderer.attach(new Surface(surfaceTexture), firstFrameLatch);
+    flutterRenderer.repaint();
+    textureId = surfaceTextureEntry.id();
+
+    super.onFlutterUiDisplayed();
+  }
+
+  @Override
+  protected void getScenarioParams(Map<String, Object> args) {
+    super.getScenarioParams(args);
+    args.put("texture_id", textureId);
+    args.put("texture_width", SURFACE_WIDTH);
+    args.put("texture_height", SURFACE_HEIGHT);
+  }
+
+  private interface SurfaceRenderer {
+    void attach(Surface surface, CountDownLatch onFirstFrame);
+
+    default boolean canProduceFrames() {
+      return true;
+    }
+
+    void repaint();
+
+    void destroy();
+  }
+
+  private static class CanvasSurfaceRenderer implements SurfaceRenderer {
+    private Surface surface;
+    private CountDownLatch onFirstFrame;
+
+    @Override
+    public void attach(Surface surface, CountDownLatch onFirstFrame) {
+      this.surface = surface;
+      this.onFirstFrame = onFirstFrame;
+    }
+
+    @Override
+    public void repaint() {
+      Canvas canvas =
+          VERSION.SDK_INT >= VERSION_CODES.M
+              ? surface.lockHardwareCanvas()
+              : surface.lockCanvas(null);
+      Paint paint = new Paint();
+      paint.setShader(
+          new LinearGradient(
+              0.0f,
+              0.0f,
+              canvas.getWidth(),
+              canvas.getHeight(),
+              new int[] {
+                0xFF00FFFF, 0xFFFF00FF, 0xFFFFFF00,
```

Comment:
Might help do add comments for the colors, so people don't have to work it out in their head and to help people know the packing order.